[Blazor] Improve auto render mode selection strategy

MackinnonBuck opened the pull request:
# Improve auto render mode selection strategy

This PR introduces an improved auto render mode selection strategy that estimates whether WebAssembly resources are already cached.

## Description

The current auto mode implementation uses a purely timeout-based mechanism for determining whether Blazor Server or Blazor WebAssembly should be used for an auto component. This PR improves the auto mode implementation to factor in whether WebAssembly resources are already cached by checking the `resources.hash` value in `blazor.boot.json`. A larger, 3 second timeout still exists in case that estimation turns out to be incorrect and Blazor WebAssembly takes too long to load.

Here's a summary of the changes in this PR:
* Added some logic in `Boot.Web.ts` that uses a hash in `localStorage` to determine if WebAssembly resources are likely cached
  * That information is used when deciding whether to start Blazor WebAssembly or Blazor Server when an auto component is encountered
* Moved the core auto mode decision making logic to `RootComponentManager` and simplified the rules:
  * For any given auto component:
    * If the WebAssembly runtime has started, use WebAssembly
    * If we've determined that WebAssembly resources would take too long to load, use Server
    * Otherwise, don't activate the component yet
  * This mostly limits the job of `Boot.Web.ts` to be determining the conditions under which to start the WebAssembly or Server runtimes
* Improved `RootComponentManager` so it can determine by itself whether an update is pending for SSR'd interactive root components 
  * The previous implementation relied on `Boot.Web.ts` calling `rootComponentManager.handleUpdatedComponentDescriptors()` everywhere
* Added auto mode E2E tests


Fixes #49580

MackinnonBuck committed:
Some auto mode tests

MackinnonBuck committed:
Use hash to detect if resources are cached

pavelsavara commented on src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:
```diff
@@ -51,8 +51,8 @@ function getValueU64(ptr: number) {
 export const monoPlatform: Platform = {
-  load: function load(options: Partial<WebAssemblyStartOptions>) {
-    return createRuntimeInstance(options);
+  load: function load(options: Partial<WebAssemblyStartOptions>, onConfigLoaded?: (loadedConfig: MonoConfig) => void) {
```
I think that the callback could be added to `WebAssemblyStartOptions`
Or is that a public type ?

MackinnonBuck replied:
Yes, `WebAssemblyStartOptions` is a public type

pavelsavara commented on src/Components/Web.JS/src/Boot.WebAssembly.Common.ts:
```diff
@@ -134,8 +147,20 @@ export async function startWebAssembly(components?: WebAssemblyComponentDescript
+
 export function loadWebAssemblyPlatform(): Promise<void> {
-  platformLoadPromise ??= monoPlatform.load(options ?? {});
+  platformLoadPromise ??= monoPlatform.load(options ?? {}, resolveBootConfigPromise);
```
Here you could have more complex method like
```js
async (config)=>{
    resolveBootConfigPromise(config)
    // delay wasm downloads by 1 second, until SSR is finished
    await new Promise(resolve => globalThis.setTimeout(resolve, 1000));
    // or 
    await untilSsrIsDone()
}
```

MackinnonBuck replied:
That's a good suggestion. That change would really only impact stream rendering, since this code wouldn't run until after the initial SSR completes.

However, I think it's actually desirable to start loading WebAssembly resources while stream rendering is happening. Interactive components get activated when stream rendering completes, so loading WebAssembly assets in parallel with the stream response could significantly reduce the time to interactivity after the stream response ends. Also, from my testing, the stream rendering experience wasn't particularly negatively impacted by the background WebAssembly download. The main scenario that I found to be problematic on slow connections was:
* An auto mode component gets SSR'd, so WebAssembly bits start downloading
* The SSR completes, but WebAssembly resources haven't finished downloading yet
* The Blazor Server SignalR connection starts up, but its speed is severely impacted by the ongoing WebAssembly download

But even in that case, I don't think there's an obvious solution that addresses all customers' needs. Some customers might want to throttle the WebAssembly download for a better Blazor Server experience, but others might want to prioritize the WebAssembly download so that it's more likely to be cached on the next visit. I think these are considerations to take into account when we implement #48756, but it might be tricky to come up with a solution in this PR that works best for the majority of cases, especially since we don't yet have data on what customers want for their apps. Since #48756 was pushed to .NET 9, I would propose that we wait until then to implement APIs to configure this. Does that sound fine to you?

pavelsavara replied:
Sure, I just wanted to hint at the possibility

mkArtakMSFT commented on src/Components/Web.JS/src/Services/RootComponentManager.ts:
```diff
@@ -30,39 +32,58 @@ type RootComponentInfo = {
+  private _didWebAssemblyFailToLoadQuickly = false;
+
+  public constructor() {
+    this.updateAllRootComponentsAfterRendererAttached(WebRendererId.Server);
+    this.updateAllRootComponentsAfterRendererAttached(WebRendererId.WebAssembly);
```
Each of these calls will wait for `rendererAttached` and then call `updateAllRootComponents`.
So technically, if there are both webassembly and server interactive root components, all will flicker twice? I am just wondering, if it will make sense to wait a little bit longer and call the `updateRootComponents` only once?

MackinnonBuck replied:
No flickering happens here. What these calls do is wait for the specific renderer to attach (which indicates that the platform is fully initialized), then call `updateAllRootComponents()`. The `updateAllRootComponents()` function looks at each root component to determine if:
* It's never been seen before, in which case it gets enabled for interactivity
* Or, it's been seen before, but it's parameters have changed, in which case send the updated parameters to .NET
* Or, it's been seen before, but now it's been removed from the page in an SSR update, so we dispose it in .NET
* Or, none of the above cases happen, in which case it's a no-op for the given component

It can kind of be thought of as calling `StateHasChanged()`, except for that it doesn't always trigger a re-render. There isn't really a downside to calling it in cases where we don't end up needing to.

mkArtakMSFT commented on src/Components/Web.JS/src/Services/RootComponentManager.ts:
```diff
@@ -99,11 +120,28 @@ export class RootComponentManager {
+      return 'server';
+    }
+
+    return null;
```
How will the `null` result be interpreted by the callers here?

MackinnonBuck replied:
The `null` indicates that a render mode could not be determined yet. This prevents the component from becoming interactive. At some point in the future (if WebAssembly starts up or fails to load quickly enough), the component will be activated via a call to `updateAllRootComponents()`.

mkArtakMSFT reviewed (approved):
Thanks @MackinnonBuck.
I sign off but please wait for at least somebody else to also review this and approve.

surayya-MS reviewed (approved):
I'm not very familiar with this area of code, but everything looks very logical and in line with the previous discussions regarding this issue. Also, great work on the e2e tests!

MackinnonBuck committed:
Improvements and cleanup

MackinnonBuck commented:
Thanks for the reviews, @surayya-MS and @mkArtakMSFT.

I've just made some further improvements. If you feel inclined to re-review, I would appreciate it. Here's what I changed:
* `Boot.Web.ts` is minimal, and logic around starting .NET runtimes and picking an auto mode has been moved to `WebRootComponentManager`
* Unified the "root component" representation with a `RootComponentManager` interface so that we didn't have two sets of logic (one for a fixed set of root components and one for a mutable set).
* Changed the auto mode logic so that WebAssembly will be used if WebAssembly resources are _loaded_, not just if the WebAssembly runtime has _started_
  * This makes it so that, in the case of stream rendering, WebAssembly gets used if WebAssembly resources finish loading by the time stream rendering completes
* A Blazor Server circuit does not get created until a Blazor Server component attempts to get activated
  * Prior to this change, the circuit would get created as soon as a Blazor Server component got discovered, even if it wasn't activated for interactivity yet

MackinnonBuck committed:
Update Boot.Server.Common.ts

surayya-MS reviewed (approved):
I think the code looks cleaner now. As for the changed logic for using WebAssembly when the resources are loaded and for creating a circuit for Server later seems logical to me.

MackinnonBuck committed:
Simplified WebAssembly component activation

MackinnonBuck committed:
Update InitialRootComponentsList.ts
